Xamarin/DrawXShared/RealmDraw.cs

```diff
@@ -0,0 +1,552 @@
+
+    However, with the paths, if we only worry about drawing _added_ or _changed_ paths
+    then we don't care if they are completed or not, just that they are dirty.
+
+    Caching and Responsiveness
+    --------------------------
+    Ideally, we want to have almost all the drawn content cached in a bitmap for redisplay, 
+    and only draw new line segments for each added point. It is relatively easy to optimise for
+    local draw updates because we know when touched that we are drawing a single added segment.
+
+    There are two distinct patterns which can cause a poor demo with synchronised Draw.
+    - A long, winding path being updated - there may be enough lag that we have to add more than 
+      one point to it locally but we don't want to redraw the whole thing from scratch.      
+    - Many small, single "dab" strokes being drawn, say from someone tapping a display, 
+      which mean we have, at least, a TouchesBegan and TouchesEnded and probably AddPoint in between.
+    
+    We make simply monitor changes - any *added* or *changed* paths get redrawn. 
+
+    Most importantly, to get the fastest possible response as the user moves their finger,
+    we draw the local line immediately as a continuation of the path they started drawing 
+    earlier. (See _currentlyDrawing)
+    
+    */
+    public class RealmDraw
+    {
+        private const float NORMALISE_TO = 4000.0f;
+        private const float PENCIL_MARGIN = 4.0f;
+        private const float INVALID_LAST_COORD = -1.0f;
+        private float _lastX = INVALID_LAST_COORD;
+        private float _lastY = INVALID_LAST_COORD;
+
+        #region Synchronised data
+        private Realm _realm;
+        private IQueryable<DrawPath> _allPaths;  // we observe all and filter based on changes
+        #endregion
+
+        #region GUI Callbacks
+        internal Action RefreshOnRealmUpdate { get; set; }
+
+        internal Action CredentialsEditor { get; set; }
+        #endregion
+
+        #region DrawingState
+        private bool _isDrawing = false;
+        private bool _ignoringTouches = false;
+        private DrawPath _drawPath;
+        private SKPath _currentlyDrawing;  // caches for responsive drawing on this device
+        private float _canvasWidth, _canvasHeight;
+        private IList<DrawPath> _pathsToDraw = null;  // set in notification callback
+        #endregion
+
+        #region CachedCanvas
+        private int _canvasSaveCount;  // from SaveLayer
+        private bool _hasSavedBitmap = false;  // separate flag so we don't rely on any given value in _canvasSaveCount
+        private bool _redrawPathsAtNextDraw = true;
+        #endregion
+
+        #region Touch Areas
+        private SKRect _loginIconRect;
+        private SKRect _loginIconTouchRect;
+        //// setup in DrawBackground
+        private float _pencilWidth;
+        private float _pencilsTop;
+        private int _numPencils;
```

Comment:
I guess it could be replaced by using the static `SwatchColor.ColorsByName.Count` a couple of times if you would rather not cache it.